removing location type staging

calellowitz opened the pull request:
@sravfeyn @gcapalbo

calellowitz committed:
removing location type staging

calellowitz commented:
tests are passing locally

sravfeyn commented on corehq/apps/locations/migrations/0012_adding_indices_for_warehouse.py:
```diff
@@ -0,0 +1,25 @@
+from django.db import migrations, models
+
+
+class Migration(migrations.Migration):
```
It might be worth moving this into a branch and PR against master itself to avoid potential migration conflicts since this PR might sit here for a long time? (unless this depends on any of pg9.6 or warehosue work)

sravfeyn commented:
Looks good, g2m

calellowitz commented on corehq/warehouse/tests/utils.py:
```diff
@@ -148,31 +138,31 @@ def create_location_records_from_tree(domain, tree, batch_id):
+#             )
 
-        _create_location_types_from_tree(domain, next_tree, location_types, batch_id)
+#         _create_location_types_from_tree(domain, next_tree, location_types, batch_id)
```
yep. that is an oversight on my part, commented them out so i could look into the logic if i missed something in the fix. will remove those

calellowitz committed:
adding user location dim sql and cleaning up tests

calellowitz committed:
Merge branch 'ce/auto-batches' into ce/warehouse-location

calellowitz committed:
resolving migration conflicts

calellowitz commented:
@gcapalbo @sravfeyn 
I cleaned up some merge conflicts and got to the feedback but also added some new work. The only new commit that needs to be reviewed should be the first. The other two are just resolving conflicts.

gcapalbo commented on corehq/warehouse/transforms/sql/user_location_dim.sql:
```diff
@@ -0,0 +1,66 @@
+(
+	SELECT
+		user_dim.id as user_dim_id,
+		json_array_elements(json_array_elements(user_staging.domain_memberships::JSON) -> 'assigned_location_ids') as location_id,
```
could there be more than 1 location id from this?

calellowitz replied:
Yeah, from my understanding (and some testing), this will expand to one row per location id

gcapalbo commented on corehq/warehouse/transforms/sql/user_location_dim.sql:
```diff
@@ -0,0 +1,66 @@
+FROM {{ user_staging }} AS user_staging
+INNER JOIN {{ user_dim }} AS user_dim
+ON user_staging.user_id = user_dim.user_id
+INNER JOIN {{ location_dim }} AS location_dim
```
these inner joins would cause records to silently drop out instead of failing the process in case there's no match
 - i made a similar comment [here](https://github.com/dimagi/commcare-hq/pull/19054#commitcomment-27235402)

calellowitz replied:
That makes a lot of sense. I can work to fix that throughout these tables. Have you noticed it elsewhere besides here and the synclog

gcapalbo replied:
i haven't seen it in any of the recent prs but it might be worth adding a task to the trello to just grep for all the joins we do and make sure they don't have this vulnerability.

adding a [balancing step](https://trello.com/c/tKhbykVj/22-relook-at-how-verified-works-for-committing-records) would also help catch these kinds of errors if they were to happen

calellowitz committed:
making inner joins left joins for insert

calellowitz commented:
@gcapalbo 
It looks like those columns were already not null so I switched these to a left join. Let me know if you have any other feedback

calellowitz committed:
changing to left joins for synclog

calellowitz committed:
more left join replacements

calellowitz commented:
I looked for all the other inner joins that feed into foreign key references and switched them to left joins. So this should all be resolved
